compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -178,10 +179,32 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 block.and(Rvalue::Use(Operand::Move(Place::from(result))))
             }
             ExprKind::Cast { source } => {
-                let source = unpack!(
-                    block =
-                        this.as_operand(block, scope, &this.thir[source], None, NeedsTemporary::No)
-                );
+                let source = &this.thir[source];
+
+                // Casting an enum to an integer is equivalent to computing the discriminant and
+                let source = if let ty::Adt(adt_def, ..) = source.ty.kind() && adt_def.is_enum() {
```

Comment:
`ty.is_enum()` is a thing, that should work here, shouldn't it?

Reply:
sure, but we need the `adt_def` inside the block